dialogue-java-client/src/main/java/com/palantir/dialogue/HttpChannel.java

```diff
@@ -84,7 +84,6 @@ public static HttpChannel of(HttpClient client, URL baseUrl, Duration requestTim
         // Fill request body and set HTTP method
         switch (endpoint.httpMethod()) {
             case GET:
-                Preconditions.checkArgument(!request.body().isPresent(), "GET endpoints must not have a request body");
```

Comment:
+1 to rfink's suggestion (co-locating the GET related validation with the actual mention of GET in this switch seems nice), could we just pop a line in here saying:

```java
if (request.body().isPresent()) {
  return Futures.immediateFailedFuture(
                    new SafeIllegalArgumentException("GET endpoint must not have a request body"));
}
```

I know java's control flow with async stuff is a bit crap (would love rust's `?` operator here), but maybe the verbosity is fine?

Reply:
This means every client must re-implement the same validation, I’d like to generalize as much as possible to avoid similar issues to our okhttp mess. happy to leave both in place here, but I think the generalized handler makes it easier to build additional client implementations with equivalent semantics.

Reply:
To rephrase; I don’t think this validation is a concern of client implementations, but a general dialogue concept.

Reply:
As an alternative, could we consider stronger types that avoid optional fields where they don't make sense semantically?

Reply:
as long as we don't have to include derive4j ;)

Reply:
Always in favor of moving validation to the type system :-)
That would be a larger break than the proposed request change, is that something we’re amenable to?

Reply:
I think it's now or never.

Reply:
Ack, before we change this I'd like to put some thought into the request body API. I put together quick netty client implementation (very, very naive at this point), but it exposes some api limitations that are worth considering before we lock things down.